cmd/influx_inspect/cardinality/aggregators/aggregators.go

```diff
@@ -0,0 +1,246 @@
+		header:   simpleHeader,
+		EstTitle: est,
+		NewNode:  func(inner bool) NodeWrapper { return NodeWrapper{newSimpleNode(inner, newCounterFn)} },
+		counter:  newCounterFn,
+	}
+}
+
+func newDetailedNodeFactory(newCounterFn func() report.Counter, est string) *RollupNodeFactory {
+	return &RollupNodeFactory{
+		header:   detailedHeader,
+		EstTitle: est,
+		NewNode:  func(isLeaf bool) NodeWrapper { return NodeWrapper{newDetailedNode(isLeaf, newCounterFn)} },
+		counter:  newCounterFn,
+	}
+}
+
+type simpleNode struct {
+	sync.Mutex
+	report.Counter
+	rollupNodeMap
+}
+
+func (s *simpleNode) Children() rollupNodeMap {
+	return s.rollupNodeMap
+}
+
+func (s *simpleNode) child(key string, isLeaf bool) NodeWrapper {
+	if s.isLeaf() {
+		panic("Trying to get the child to a leaf node")
+	}
+	s.Lock()
+	defer s.Unlock()
+	c, ok := s.Children()[key]
+	if !ok {
+		c = nodeFactory.NewNode(isLeaf)
+		s.Children()[key] = c
+	}
+	return NodeWrapper{c}
+}
+
+func (s *simpleNode) isLeaf() bool {
+	return s.Children() == nil
+}
+
+func newSimpleNode(isLeaf bool, fn func() report.Counter) *simpleNode {
+	s := &simpleNode{Counter: fn()}
+	if !isLeaf {
+		s.rollupNodeMap = make(rollupNodeMap)
+	} else {
+		s.rollupNodeMap = nil
+	}
+	return s
+}
+
+func (s *simpleNode) RecordSeries(db, rp, _ string, key, _ []byte, _ models.Tags) {
+	s.Lock()
+	defer s.Unlock()
+	s.recordSeriesNoLock(db, rp, key)
+}
+
+func (s *simpleNode) recordSeriesNoLock(db, rp string, key []byte) {
+	s.Add([]byte(fmt.Sprintf("%s.%s.%s", db, rp, key)))
+}
+
+func (s *simpleNode) Print(tw *tabwriter.Writer, _ bool, db, rp, ms string) error {
+	_, err := fmt.Fprintf(tw, "%s\t%s\t%s\t%d\n",
+		db,
+		rp,
+		ms,
+		s.Count())
+	return err
+}
+
+type detailedNode struct {
+	simpleNode
+	fields report.Counter
+	tags   map[string]report.Counter
+}
+
+func newDetailedNode(isLeaf bool, fn func() report.Counter) *detailedNode {
+	d := &detailedNode{
+		simpleNode: simpleNode{
+			Counter: fn(),
+		},
+		fields: fn(),
+		tags:   make(map[string]report.Counter),
+	}
+	if !isLeaf {
+		d.simpleNode.rollupNodeMap = make(rollupNodeMap)
+	} else {
+		d.simpleNode.rollupNodeMap = nil
+	}
+	return d
+}
+
+func (d *detailedNode) RecordSeries(db, rp, ms string, key, field []byte, tags models.Tags) {
+	d.Lock()
+	defer d.Unlock()
+	d.simpleNode.recordSeriesNoLock(db, rp, key)
+	d.fields.Add([]byte(fmt.Sprintf("%s.%s.%s.%s", db, rp, ms, field)))
+	for _, t := range tags {
+		// Add database, retention policy, and measurement
+		// to correctly aggregate in inner (non-leaf) nodes
+		canonTag := fmt.Sprintf("%s.%s.%s.%s", db, rp, ms, t.Key)
+		tc, ok := d.tags[canonTag]
+		if !ok {
+			tc = nodeFactory.counter()
+			d.tags[canonTag] = tc
+		}
+		tc.Add(t.Value)
+	}
+}
+
+func (d *detailedNode) Print(tw *tabwriter.Writer, printTags bool, db, rp, ms string) error {
+	seriesN := d.Count()
+	fieldsN := d.fields.Count()
+	var tagKeys []string
+	tagN := uint64(0)
+
+	if d.isLeaf() {
+		tagKeys = make([]string, 0, len(d.tags))
```

Comment:
Might be more natural here to check `printTags` instead of `isLeaf`.

In general I don't think `isLeaf` is adding much - a few safety checks that I don't think are needed, and also whether we create a `map` or not for the leaves.

Reply:
We can probably just remove `isLeaf()` and the `isLeaf` argument for node creation?

Reply:
But I won't hold the PR on this.

Reply:
I dislike the empty map memory wastage, TBH.  Agree that the check above should be `printTags`